docs/developer/policy-engine.md

```diff
@@ -0,0 +1,248 @@
+# Policy Engine
+
+## Register policy functions
+
+The following will document how to register a constraint function to the policy engine.
+
+### Step 1: Create a module
+
+As an example, we create the module `:extensions:policy:ids-policy`. Add a `build.gradle.kts` file:
+
+```
+val infoModelVersion: String by project
+val rsApi: String by project
+
+plugins {
+    `java-library`
+}
+
+
+dependencies {
+    api(project(":spi"))
+    implementation(project(":data-protocols:ids:ids-spi"))
+}
+```
+
+To ensure that the extension will
+be loaded, add it to the `settings.gradle.kts`:
+
+```
+include(":extensions:policy:ids-policy")
+```
+
+### Step 2: Implement a `ServiceExtension`
+
+The new policy extension will provide a service extension class and two constraint functions. In
+`IdsPolicyExtension`, first, the scope `ALL_SCOPES` is bound to the rule type `USE`. Next, both policy 
+functions are registered to the policy engine.
+
+```java
+public class IdsPolicyExtension implements ServiceExtension {
+
+    private static final String ABS_SPATIAL_POSITION = "ids:absoluteSpatialPosition";
+    private static final String PARTNER_LEVEL = "ids:partnerLevel";
+
+    @Inject
+    private RuleBindingRegistry ruleBindingRegistry;
+
+    @Inject
+    private PolicyEngine policyEngine;
+
+    @Override
+    public String name() {
+        return "IDS Policy";
+    }
+
+    @Override
+    public void initialize(ServiceExtensionContext context) {
+        ruleBindingRegistry.bind(USE, ALL_SCOPES);
+
+        policyEngine.registerFunction(ALL_SCOPES, Permission.class, ABS_SPATIAL_POSITION, new AbsSpatialPositionConstraintFunction());
+        policyEngine.registerFunction(ALL_SCOPES, Permission.class, PARTNER_LEVEL, new PartnerLevelConstraintFunction());
+    }
+}
+```
+
+### Step 3: Implement an `AtomicConstraintFunction`
+
+#### Absolute Spatial Position Constraint Function
+
+```java
+public class AbsSpatialPositionConstraintFunction implements AtomicConstraintFunction<Permission> {
+    
+    @Override
+    public boolean evaluate(Operator operator, Object rightValue, Permission rule, PolicyContext context) {
+        var region = context.getParticipantAgent().getClaims().get("region");
+        switch (operator) {
+            case EQ:
+                return Objects.equals(region, rightValue);
+            case NEQ:
+                return !Objects.equals(region, rightValue);
+            case IN:
+                return ((Collection<?>) rightValue).contains(region);
+            default:
+                return false;
+        }
+    }
+}
+```
+
+```java
+class AbsSpatialPositionConstraintFunctionTest {
```

Comment:
I wouldn't include tests in this document, as it'll be up to the developer to write them in the preferred "flavour"